File: math/leetcode_Plus_One.java
/*Plus_One

Given a number represented as an array of digits, plus one to the number.
*/

public class leetcode_Plus_One {

    public static void main(String[] args) {

    }

    public static int[] plusOne(int[] digits) {
        // Take care of the last one.
        int carry = (digits[digits.length - 1] + 1 >= 10) ? 1 : 0;
        digits[digits.length - 1] = (digits[digits.length - 1] + 1) % 10;
        for (int i = digits.length - 2; i >= 0; --i) {
            int digit = digits[i];
            digits[i] = (digit + carry) % 10;
            carry = (digit + carry >= 10) ? 1 : 0;
        }

        // After iteration, check if carry equals 1.
        if (carry == 1) {
            int[] res = new int[digits.length + 1];
            res[0] = 1;
            for (int i = 0; i < digits.length; ++i) {
                res[i + 1] = digits[i];
            }
            return res;
        }
        return digits;
    }
}

File: math/leetcode_Plus_One.cpp
/**
 * Copyright 2016
 * author: madarfacar
 *
 * leetcode plus one
 */

#include <stdio.h>
#include <vector>

using std::vector;


class Solution {
 public:
  vector<int> plusOne(vector<int>& digits) {
    bool carry = false;

    for (int i = digits.size() - 1; i >= 0; --i) {
      int digit = digits[i] + (i == digits.size() - 1 ? 1 : 0) + (carry ? 1 : 0);
      digits[i] = digit % 10;
      carry = digit >= 10;
    }

    if (carry) {
      digits.insert(digits.begin(), 1);
    }

    return digits;
  }
};


int main() {
  Solution sol;
  vector<int> digits {9, 9, 9, 9};
  vector<int> res = sol.plusOne(digits);

  for (int val : res) {
    printf("%d ", val);
  }
  printf("\n");
}
